1.x: new operators buffer(While/Until) with predicate-based boundary

akarnokd opened the pull request:
The need for a buffer operator that splits the source sequence into continuous non-overlapping buffers based on the values in the sequence comes up from time to time. For, me, it reached the point where I'd consider adding operator(s) with this functionality into RxJava.

There are two variants of the required functionality: split before the element is added to the buffer: `bufferWhile`; and split after the element has been added to the buffer: `bufferUntil`.  Both cases, the predicate has to return `true` to trigger a split and emission of the current buffer.

I've added 2-2 overloads of these methods to `Observable` where the second overloads allow specifying the expected capacity use of the buffers (for performance reasons). These operators don't support backpressure and go unbounded, just like `buffer(Observable)` and others.

In the operator itself, I've added backpressured versions of `bufferWhile` and `bufferUntil`, which use a prefetch queue and a request-dependent queue-drain approach. It is up for a discussion what to do with these (provided this PR is of interest at all): 1) use them as the basis for the methods instead, 2) expose them through a different name and 3) drop them.

stevegury commented on src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java:
```diff
@@ -0,0 +1,580 @@
+            for (;;) {
+
+                long r = requested.get();
+                long e = 0L;
```
For readability would you mind giving longer names to those variable?

stevegury commented on src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java:
```diff
@@ -0,0 +1,580 @@
+                return;
+            }
+            
+            if (emit) {
```
I think it's less confusing to not emit the buffer when it's null.
I'd add a check for `! b.isEmpty()` in the if.

stevegury commented:
w.r.t. your question, I would prefer having the backpressure-aware operator be the default implementation.

akarnokd commented:
Updated:
- Using backpressure and a standard prefetch of RxRingBuffer.SIZE
- BufferWhile won't emit an empty initial buffer in case the first value triggers a new buffer
- Added `@Experimental` tags
- Added backpressure-tests and and explicit tests with empty or error sources.
- renamed local variables

stevegury commented:
:+1:

akarnokd commented:
/cc @davidmoten @artem-zinnatullin

artem-zinnatullin commented:
LGTM except operators names.

* `bufferWhile(func)` sounds like it will buffer *while* you're returning `true` from `func` and it'll stop_buffering/will_split when you'll return `false`.
* `bufferUntil(func)` sounds like it will buffer *until* you're returning `false` and it'll stop_buffering/will_split when you'll return `true`.

Maybe something like:

* `bufferAndSplitWhen(func)`
* `bufferAndSplitAfter(func)`

not a great names too, but at least more obvious (to me).

Ping others: what do you think about operators names?

stevegury commented:
I prefer the `bufferWhile` `bufferUntil` version, it's idiomatic in functional languages like Scala and other to use those words.

vanniktech commented on src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java:
```diff
@@ -0,0 +1,487 @@
+            this.capacityHint = capacityHint;
+            this.predicate = predicate;
+            Queue<Object> q;
+            if (prefetch >= 10 * 1000 * 1000) {
```
What about `TimeUnit.Milliseconds.toNanos(10)`?

artem-zinnatullin replied:
It's not time, it's prefetch amount. Also, your variant will require runtime calculation, David's may be calculated at compile time (nano win!).

akarnokd replied:
Doesn't matter, I'll remove it.

artem-zinnatullin commented:
Well, for me it sounds like `takeWhile()`/`takeUntil()` and `skipWhile()`/`skipUntil()`.

Compare:

```java
Observable
  .range(1, 1000000)
  .bufferWhile(number -> number % 10 == 0)
  .subscribe(numbers -> System.out.println(numbers));

// Here I'll think that it'll buffer numbers by some unknown mechanism 
// and then stop on first number that % 10 == 0. 
// Subscriber will receive not complete set of items.
```

And:

```java
Observable
  .range(1, 1000000)
  .bufferAndSplitWhen(number -> number % 10 == 0)
  .subscribe(numbers -> System.out.println(numbers));

// Here I clearly see that subscriber will receive all items.
// And that lambda specifies split points, not the end of the stream.
```

idk about others but I'll have to read javadoc each time to understand what to expect with current operators names.

davidmoten commented on src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java:
```diff
@@ -0,0 +1,487 @@
+                    }
+                    
+                    if (emit) {
+                        localSubscriber.onNext(localBuffer);
```
As predicate.call could take a non-trivial time we could check for unsubscription before emitting. It's a trade-off between checking the volatile unsubscribed flag and reducing work after unsubscription but in other operators we normally do the unsubscribed check.

artem-zinnatullin replied:
Not sure it worth it, passing every item to a predicate function already slows down the emission… `SafeSubscriber` will check subscriber state for us anyway.

Subscription state checked in the beginning of each iteration of the cycle ^.

davidmoten replied:
>Not sure worth it
Hard to say but it's a convention throughout the codebase at the moment. @akarnokd your opinion (being such a performance oriented fellow)?

akarnokd replied:
The most likely unsubscription will happen just after emission (take). There will be always a window between calling unsubscribe() and receiving an event. So I don't think there is a need for more eagerness than what's in the drain loop now. As for `SafeSubscribe`, last time I checked it didn't do hard-gating on `isUnsubscribed` but if one is really eager, his/her `onNext` of `Subscriber` can have extra checks before acting on the value.

davidmoten commented:
I'll get back to reviewing the code in a bit but in terms of names an alternative is `toListWhile` and `toListUntil` that could also exist as aliases to the `buffer` versions. This is the api method I went with in *rxjava-extras* that implements this functionality ([Transformers.java](https://github.com/davidmoten/rxjava-extras/blob/a4edc924477b063586cee00be96412db2f444f95/src/main/java/com/github/davidmoten/rx/Transformers.java)). I had a dedicated operator for this stuff till I knocked up `Transformers.stateMachine` (discussion in #3442 that hasn't got any interest seemingly) and `toListWhile` and `toListUntil` goes through that (via `collectWhile/Until`).  I'll be interested to compare performance with the dedicated operator by @akarnokd in this PR. 

A natural extension to this PR that could be incorporated now is using `collectWhile` and `collectUntil` which `bufferWhile` and `bufferUntil` would be built on top of so that we aren't just list focused but can use arbitrary collection types.

akarnokd commented:
I could rename it to `bufferSplit`, change the parameters to include an isBefore boolean and only have 2 overloads.

akarnokd commented:
Updated the PR:

- removed that magic number, wasn't really relevant in the current setup
- combined the operators into `bufferSplitWhen` and an `checkAfter` flag to express the time the predicate should be checked.

akarnokd commented:
`testErrorThrownIssue1685` keeps failing for some reason on CI (but never locally).

stevegury commented:
:+1: 
I also noticed recently that `testErrorThrownIssue1685` started failing more frequently on TravisCI.

artem-zinnatullin commented on src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java:
```diff
@@ -0,0 +1,408 @@
+        return parent;
+    }
+    
+    static abstract class BoundedSubscriber<T> extends Subscriber<T> {
```
nice abstraction :+1:

artem-zinnatullin commented on src/test/java/rx/internal/operators/OperatorBufferPredicateBoundaryTest.java:
```diff
@@ -0,0 +1,473 @@
+public class OperatorBufferPredicateBoundaryTest {
+
+    @Test
+    public void bufferWhilePredicateNull() {
```
Pls fix tests names

akarnokd replied:
It helps me tell the two modes apart.

artem-zinnatullin replied:
k, since they live in dedicated clas, it's ok

artem-zinnatullin commented:
:+1:

davidmoten commented:
I see that the predicate is based only on the latest emission from upstream. If the predicate was a `Func2<List<T>,T, Boolean>` then the operator would even more useful and with little change to the existing operator.

davidmoten commented:
I prefer `bufferUntil/While` over `bufferSplit(predicate, true/false)`. The second version is not as readable as the first.

stevegury commented:
+1 on `bufferUntil`/`bufferWhile` but I'm not strongly opinionated about that.

akarnokd commented:
What's the verdict? bufferWile/bufferUntil where until's function has to emit false?

akarnokd commented:
Restored the original naming. In both operators, the predicate has to return true to trigger the next buffer. Let me know if you want some other rule.

stevegury commented:
:+1: I'm with @davidmoten about the naming.
Ping @stealthcode @abersnaze for their input.

ZacSweers commented:
👍 . Could see this having a lot of applications for something like batching and analytics when you want to wait for certain conditions.
